gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/converter/EnvelopeSchemaConverter.java

```diff
@@ -1,152 +1,145 @@
    */
-  @Override
-  public String convertSchema(Schema inputSchema, WorkUnitState workUnit) throws SchemaConversionException {
-    return EnvelopeSchemaConverter.class.getName();
+  protected Schema getPayloadSchema(GenericRecord inputRecord)
+      throws Exception {
+    Optional<Object> schemaIdValue = AvroUtils.getFieldValue(inputRecord, payloadSchemaIdField);
+    if (!schemaIdValue.isPresent()) {
+      throw new Exception("Schema id with key " + payloadSchemaIdField + " not found in the record");
+    }
+    String schemaKey = String.valueOf(schemaIdValue.get());
+    return schemaCache.get(schemaKey);
   }
 
   /**
-   * Get actual schema from registry and deserialize payload using it.
+   * Get payload field and convert to byte array
+   *
+   * @param inputRecord the input record which has the payload
+   * @return the byte array of the payload in the input record
    */
-  @Override
-  public Iterable<GenericRecord> convertRecord(String outputSchema, GenericRecord inputRecord, WorkUnitState workUnit)
-      throws DataConversionException {
-    try {
-      String schemaIdField = workUnit.contains(PAYLOAD_SCHEMA_ID_FIELD) ?
-          workUnit.getProp(PAYLOAD_SCHEMA_ID_FIELD) : DEFAULT_PAYLOAD_SCHEMA_ID_FIELD;
-      String payloadField = workUnit.contains(PAYLOAD_FIELD) ?
-          workUnit.getProp(PAYLOAD_FIELD) : DEFAULT_PAYLOAD_FIELD;
-      String schemaKey = String.valueOf(inputRecord.get(schemaIdField));
-      Schema payloadSchema = (Schema) this.registry.getSchemaByKey(schemaKey);
-      byte[] payload = getPayload(inputRecord, payloadField);
-      GenericRecord outputRecord = deserializePayload(payload, payloadSchema);
-      if (this.fieldRemover.isPresent()) {
-        payloadSchema = this.fieldRemover.get().removeFields(payloadSchema);
-      }
-      return new SingleRecordIterable<>(AvroUtils.convertRecordSchema(outputRecord, payloadSchema));
-    } catch (IOException | SchemaRegistryException | ExecutionException e) {
-      throw new DataConversionException(e);
+  protected byte[] getPayloadBytes(GenericRecord inputRecord) throws Exception {
+    Optional<Object> payloadValue = AvroUtils.getFieldValue(inputRecord, payloadField);
+    if (!payloadValue.isPresent()) {
+      throw new Exception("Payload with key " + payloadField + " not found in the record");
     }
-  }
 
-  /**
-   * Get payload field from GenericRecord and convert to byte array
-   */
-  public byte[] getPayload(GenericRecord inputRecord, String payloadFieldName) {
-    ByteBuffer bb = (ByteBuffer) inputRecord.get(payloadFieldName);
-    byte[] payloadBytes;
+    ByteBuffer bb = (ByteBuffer) payloadValue.get();
     if (bb.hasArray()) {
-      payloadBytes = bb.array();
+      return bb.array();
     } else {
-      payloadBytes = new byte[bb.remaining()];
+      byte[] payloadBytes = new byte[bb.remaining()];
       bb.get(payloadBytes);
+      String hexString = new String(payloadBytes, StandardCharsets.UTF_8);
```

Comment:
I got the logic in the previous implementation. So far, we don't have a use case for this. It's unclear what the right logic is.

Reply:
Can you check whether the payload is a hex-encoded string? It did look that way in one of the examples.

Reply:
In the example, it's not a hex-encoded string. In fact, it can be any format. I'll simply return the payloadBytes as the default implementation and let the subclass overwrite the behavior if needed.